### ISSUE TITLE
PROTEUS unit tests

### ISSUE THREAD

**Issue by timlichtenberg** [FormingWorlds/PROTEUS#268]
We should aim to get the test coverage to 100% before the PROTEUS method paper in JOSS. Alternatively we need to label (or remove) parts of the code that are not yet necessary anymore.

**Comment by nichollsh** [comment]
Here's a breakdown of the current test coverage, for reference.

```
Name                                         Stmts   Miss Branch BrPart  Cover
------------------------------------------------------------------------------
src/proteus/__init__.py                          4      0      0      0   100%
src/proteus/atmos_clim/__init__.py               3      0      0      0   100%
src/proteus/atmos_clim/agni.py                 199    199     48      0     0%
src/proteus/atmos_clim/common.py                49      2     12      2    93%
src/proteus/atmos_clim/dummy.py                 70     18     14      4    69%
src/proteus/atmos_clim/janus.py                137     29     44     12    72%
src/proteus/atmos_clim/wrapper.py               88     39     30      8    55%
src/proteus/cli.py                              82     82      8      0     0%
src/proteus/config/__init__.py                  13      0      0      0   100%
src/proteus/config/_atmos_clim.py               62      5      6      3    88%
src/proteus/config/_config.py                   33      0      0      0   100%
src/proteus/config/_converters.py               12      0      6      0   100%
src/proteus/config/_delivery.py                 28      0      0      0   100%
src/proteus/config/_escape.py                   17      0      0      0   100%
src/proteus/config/_interior.py                 28      0      0      0   100%
src/proteus/config/_orbit.py                    11      0      0      0   100%
src/proteus/config/_outgas.py                   24      0      0      0   100%
src/proteus/config/_params.py                   69      1      2      1    97%
src/proteus/config/_star.py                     21      0      0      0   100%
src/proteus/config/_struct.py                    9      0      0      0   100%
src/proteus/escape/wrapper.py                   48      3     20      3    91%
src/proteus/interior/__init__.py                 3      0      0      0   100%
src/proteus/interior/aragog.py                  86      1      8      1    98%
src/proteus/interior/dummy.py                   53      2     10      2    94%
src/proteus/interior/spider.py                 259    259     52      0     0%
src/proteus/interior/timestep.py                51      5     14      3    88%
src/proteus/interior/wrapper.py                 94     18     38      9    77%
src/proteus/orbit/__init__.py                    2      0      0      0   100%
src/proteus/orbit/wrapper.py                    11      0      0      0   100%
src/proteus/outgas/__init__.py                   3      0      0      0   100%
src/proteus/outgas/atmodeller.py                 6      6      2      0     0%
src/proteus/outgas/calliope.py                  52      7     22      5    84%
src/proteus/outgas/wrapper.py                   22      3     12      3    82%
src/proteus/plot/__init__.py                    22      0      0      0   100%
src/proteus/plot/_cpl_helpers.py                 7      0      0      0   100%
src/proteus/plot/cpl_atmosphere.py              51     10      8      3    78%
src/proteus/plot/cpl_atmosphere_cbar.py         72     55     14      2    22%
src/proteus/plot/cpl_elements.py                50      8     10      3    82%
src/proteus/plot/cpl_emission.py                86     21     16      5    71%
src/proteus/plot/cpl_escape.py                  51      9     12      5    78%
src/proteus/plot/cpl_fluxes_atmosphere.py       52      5      4      2    88%
src/proteus/plot/cpl_fluxes_global.py           55      9      6      3    80%
src/proteus/plot/cpl_global.py                 134     10     42      3    91%
src/proteus/plot/cpl_heatingrates.py            40     30      6      2    26%
src/proteus/plot/cpl_interior.py               137     35     46     13    64%
src/proteus/plot/cpl_interior_cmesh.py         124     28     44     12    71%
src/proteus/plot/cpl_observables.py             42      8      6      3    77%
src/proteus/plot/cpl_offchem_grid_cross.py     126    112     48      2     9%
src/proteus/plot/cpl_offchem_species.py         79     64     24      2    17%
src/proteus/plot/cpl_offchem_time.py            91     78     34      2    12%
src/proteus/plot/cpl_offchem_year.py            64     51     22      2    17%
src/proteus/plot/cpl_population.py             110     12     16      5    87%
src/proteus/plot/cpl_sflux.py                   96     23     28     10    73%
src/proteus/plot/cpl_sflux_cross.py             92     22     26      6    75%
src/proteus/plot/cpl_structure.py              109     19     18      5    80%
src/proteus/proteus.py                         231     57     66     12    75%
src/proteus/star/__init__.py                     2      0      0      0   100%
src/proteus/star/dummy.py                       30      0      6      2    94%
src/proteus/star/wrapper.py                    110     13     34      8    80%
src/proteus/utils/constants.py                  39      0      0      0   100%
src/proteus/utils/coupler.py                   275     23     82     19    88%
src/proteus/utils/data.py                      138     74     36      6    44%
src/proteus/utils/helper.py                    134     55     68     11    50%
src/proteus/utils/logs.py                       74     25     20      6    59%
src/proteus/utils/phys.py                       16      0      0      0   100%
src/proteus/utils/plot.py                      123     63     32      4    46%
src/proteus/utils/plot_offchem.py               85     76     32      0     8%
------------------------------------------------------------------------------
TOTAL                                         4596   1674   1154    214    60%

```

**Comment by nichollsh** [comment]
A lot of it is the offline chemistry stuff, which is going to be completely reworked anyway. Some of it is the AGNI wrapper, which we could address quite easily. Some of it is from the SPIDER wrapper, which would be harder to do.

**Comment by timlichtenberg** [comment]
Ok, I believe the offline chemistry lines will probably be touched anyway within the next few months I presume, so that seems doable. The SPIDER wrapper in itself is probably reasonable to touch, but I agree we shouldn't go overboard with this.

**Comment by lsoucasse** [comment]
I agree it is important to increase the coverage but I don't think the 100% coverage is a reasonable goal. It is not required for JOSS.

**Comment by timlichtenberg** [comment]
It is not required, but why is it in general unreasonable? I understand that this means rigorously sorting the current code lines. But in truth, any thing below 100% will in the long run become much harder to keep up because in practice it will be difficult to keep track of where tests are missing, when many users are adding code lines.

**Comment by lsoucasse** [comment]
What I mean is that 100% coverage is great but it requires a lot of effort and doesn't mean that it is bug-free because it depends on how to test things.
The mors bug is a good example, everything works fine at low time evolution and then you add one data point at a bigger time (which does not increase the coverage) and then the test fails.
Also at the moment we change the test data more or less at each PR because adding new physics/parameters change everything. We are happy because the test pass but it does not guarantee the implementation is correct.
In conclusion, I definitely agree we need to increase the coverage but it should not be the unique focus point.

**Comment by stefsmeets** [comment]
I agree with @lsoucasse that 100% is an impossible target. I think that 80% is a solid and realistic target for this code base. Just think about it in this way, to achieve 100% coverage, every single branch (e.g. `if` statement) must be tested.

+1 on @lsoucasse point to consider is test quality. 100% test coverage does not mean 100% correctness. It just means that the code runs and does not crash.

**Comment by timlichtenberg** [comment]
Ok ok :) Let's have 70-80% as a continuous target then. Testing that the code does not crash is a already a massively helpful goal to achieve.

**Comment by nichollsh** [comment]
Current code coverage is ~67%. This is the current breakdown, sorted by  the number of missed calls.

```
Name                                        Stmts   Miss Branch BrPart  Cover
-----------------------------------------------------------------------------
src/proteus/__init__.py                         4      0      0      0   100%
src/proteus/atmos_clim/__init__.py              3      0      0      0   100%
src/proteus/config/__init__.py                 13      0      0      0   100%
src/proteus/config/_delivery.py                38      0      0      0   100%
src/proteus/config/_observe.py                  7      0      0      0   100%
src/proteus/config/_outgas.py                  26      0      0      0   100%
src/proteus/interior/__init__.py                3      0      0      0   100%
src/proteus/observe/__init__.py                 2      0      0      0   100%
src/proteus/orbit/__init__.py                   2      0      0      0   100%
src/proteus/outgas/__init__.py                  3      0      0      0   100%
src/proteus/plot/__init__.py                   17      0      0      0   100%
src/proteus/plot/_cpl_helpers.py                7      0      0      0   100%
src/proteus/star/__init__.py                    2      0      0      0   100%
src/proteus/star/dummy.py                      30      0      6      2    94%
src/proteus/utils/constants.py                 44      0      0      0   100%
src/proteus/utils/phys.py                      16      0      0      0   100%
src/proteus/config/_params.py                  69      1      2      1    97%
src/proteus/config/_converters.py              15      2      6      0    90%
src/proteus/config/_orbit.py                   27      2      4      2    87%
src/proteus/interior/dummy.py                  67      2     12      2    95%
src/proteus/atmos_chem/__init__.py              3      3      0      0     0%
src/proteus/config/_escape.py                  33      3     10      3    86%
src/proteus/orbit/dummy.py                     19      3     10      2    76%
src/proteus/outgas/wrapper.py                  23      3     12      3    83%
src/proteus/config/_config.py                  44      4      6      2    84%
src/proteus/atmos_chem/wrapper.py               5      5      0      0     0%
src/proteus/config/_interior.py                66      5     16      5    88%
src/proteus/observe/common.py                   6      6      0      0     0%
src/proteus/outgas/atmodeller.py                6      6      2      0     0%
src/proteus/plot/cpl_fluxes_atmosphere.py      62      7      8      4    84%
src/proteus/config/_star.py                    46      8     20      8    76%
src/proteus/plot/cpl_bolometry.py              42      8      6      3    77%
src/proteus/plot/cpl_escape.py                 72      8     10      3    87%
src/proteus/atmos_clim/common.py               78      9     20      4    85%
src/proteus/config/_struct.py                  35      9     20      7    67%
src/proteus/escape/wrapper.py                  54      9     26      4    79%
src/proteus/observe/wrapper.py                 16      9      6      1    36%
src/proteus/plot/cpl_fluxes_global.py          57      9      6      3    81%
src/proteus/plot/cpl_atmosphere.py             57     10      8      3    80%
src/proteus/plot/cpl_global.py                148     10     48      3    92%
src/proteus/orbit/wrapper.py                   67     11     20      7    77%
src/proteus/plot/cpl_population.py            119     12     16      5    87%
src/proteus/interior/aragog.py                134     14     24      8    85%
src/proteus/config/_atmos_clim.py              90     16     26      9    75%
src/proteus/star/wrapper.py                   131     16     42     10    79%
src/proteus/proteus.py                        196     18     52      6    88%
src/proteus/plot/cpl_structure.py             109     19     18      5    80%
src/proteus/atmos_clim/dummy.py                71     20     14      5    66%
src/proteus/interior/common.py                 88     20     22      4    71%
src/proteus/plot/cpl_sflux.py                  94     20     28      8    77%
src/proteus/utils/logs.py                      74     20     20      5    65%
src/proteus/plot/cpl_sflux_cross.py            90     22     26      7    73%
src/proteus/plot/cpl_emission.py               89     23     18      6    69%
src/proteus/utils/terminate.py                123     23     44     10    79%
src/proteus/plot/cpl_interior_cmesh.py        123     27     42     11    72%
src/proteus/atmos_clim/wrapper.py              94     28     30      9    70%
src/proteus/interior/wrapper.py               138     29     56     14    76%
src/proteus/outgas/calliope.py                 98     29     44     13    68%
src/proteus/utils/coupler.py                  347     29    112     21    89%
src/proteus/atmos_clim/janus.py               142     31     46     13    71%
src/proteus/interior/timestep.py              107     34     36     10    64%
src/proteus/utils/plot.py                      91     35     22      2    62%
src/proteus/plot/cpl_interior.py              160     37     50     14    66%
src/proteus/orbit/lovepy.py                    40     40      6      0     0%
src/proteus/plot/cpl_atmosphere_cbar.py        57     40      6      2    30%
src/proteus/plot/cpl_spectra.py                62     48     18      2    20%
src/proteus/utils/helper.py                   140     57     72     11    51%
src/proteus/doctor.py                          68     68     14      0     0%
src/proteus/atmos_clim/agni.py                236     70     64     17    66%
src/proteus/utils/data.py                     149     76     38      6    47%
src/proteus/observe/platon.py                  87     87     12      0     0%
src/proteus/cli.py                             92     92     12      0     0%
src/proteus/interior/spider.py                291    291     64      0     0%
-----------------------------------------------------------------------------
TOTAL                                        5264   1543   1378    305    67%
```

**Comment by nichollsh** [comment]
Main things we are missing:
1. Anything to do with SPIDER
2. Synthetic observations module 
3. PROTEUS command line interface
4. Tidal heating module

Addressing 3 and 4 should be fairly easy. I estimate that if we tested these components entirely, the coverage would be 74%. 

Addressing 1 is hard, because of PETSc. Addressing 2 is possible, and should be done, but the main barrier is downloading the 10 GB lookup data for Platon.

**Comment by timlichtenberg** [comment]
3. & 4. sounds great to me. I would not do 1., rather I would aim to retire SPIDER once we have ~2-3 papers out with Aragog in the next ~12 months. I am not entirely clear about the use of Platon, what are the unique features of it and is there not sth more lightweight? In principle Icould AGNI could fulfill the same role in the mid-term?

**Comment by nichollsh** [comment]
I have a PR addressing 3 and 4 here: #364 which is ready to be reviewed. It brings the coverage to 72%, which finally brings us to the 70-80% target that you set.

**Comment by nichollsh** [comment]
The main utility of platon is calculating transmission spectra with the correct geometry. We cannot calculate transit depth with AGNI because it is formulated with a zenith column, rather than a limb-viewing column.

It also offers higher resolution because the radiative transfer is primarily done through opacity tables, rather than with correlated-k.

**Comment by timlichtenberg** [comment]
Ok, and have you checked if both pRT and TauRex 3 have similar data requirements?

**Comment by nichollsh** [comment]
pRT is a good idea - we should definitely look into that, but it has a Python interface that would work well within PROTEUS. I have never tried TauRex hands-on.
